### PR TITLE
Upgrade MySQL Connector to 8.0.22

### PR DESCRIPTION
**Pull request by gsmet**
AsyncProviderFactory is gone entirely.
I haven't upgraded MySQL itself as it seems Docker images are not
available yet or it hasn't been released yet.

### CODE REVIEW

**Review by Sanne — APPROVED** [review]

**Review by machi1990 — APPROVED** [review]
LGTM. 

I am wondering if we should introduce a property for this feature https://github.com/mysql/mysql-connector-j/commit/6a56e51480a58b4b164ae9e3649ecc1e6ee11a69

I think this property was introduced to make native image generation work out of the box https://github.com/mysql/mysql-connector-j/pull/41 since it was failing since cleaning up abandoned connection launched a thread on a static block GraalVM is not happy about that.

We had to substitute it on our side:
https://github.com/quarkusio/quarkus/blob/master/extensions/jdbc/jdbc-mysql/runtime/src/main/java/io/quarkus/jdbc/mysql/runtime/graal/AbandonedConnectionCleanupThreadSubstitutions.java

**Comment by Sanne** [comment]
> LGTM.
> 
> I am wondering if we should introduce a property for this feature [mysql/mysql-connector-j@6a56e51](https://github.com/mysql/mysql-connector-j/commit/6a56e51480a58b4b164ae9e3649ecc1e6ee11a69)

very glad to see that! I asked that some time ago, didn't realize someone had actually contributed a fix.

**Comment by Sanne** [comment]
@machi1990 yes, we should definitely use that property. Not only it allows to remove the substitution, it will also make sure that native code and JVM code will have exactly the same semantics, which is always desireable. Are you able to volunteer?

**Comment by machi1990** [comment]
> > LGTM.
> > I am wondering if we should introduce a property for this feature [mysql/mysql-connector-j@6a56e51](https://github.com/mysql/mysql-connector-j/commit/6a56e51480a58b4b164ae9e3649ecc1e6ee11a69)
> 
> very glad to see that! I asked that some time ago, didn't realize someone had actually contributed a fix.

Yes, I think in our case we can turn this on and remove the cleanup substitution logic all together since Agroal handles that already IIRC. WDYT? @Sanne

**Comment by machi1990** [comment]
> @machi1990 yes, we should definitely use that property. Not only it allows to remove the substitution, it will also make sure that native code and JVM code will have exactly the same semantics, which is always desireable. Are you able to volunteer?

Did not realize we commented about the same time :) I can do that in a followup PR sometime this week.

**Comment by Sanne** [comment]
ah, and too bad that this wasn't declared `final` ..
 - https://github.com/mysql/mysql-connector-j/commit/6a56e51480a58b4b164ae9e3649ecc1e6ee11a69#diff-cacd06ef2d6fd561bc89b568a3dac43fdd37978aa1419b67fb0ad8f55fce503aR57

I hope it won't be a problem.

**Comment by machi1990** [comment]
> ah, and too bad that this wasn't declared `final` ..
> 
> * [mysql/mysql-connector-j@6a56e51#diff-cacd06ef2d6fd561bc89b568a3dac43fdd37978aa1419b67fb0ad8f55fce503aR57](https://github.com/mysql/mysql-connector-j/commit/6a56e51480a58b4b164ae9e3649ecc1e6ee11a69#diff-cacd06ef2d6fd561bc89b568a3dac43fdd37978aa1419b67fb0ad8f55fce503aR57)
> 
> I hope it won't be a problem.

@gsmet @Sanne I pushed a commit https://github.com/quarkusio/quarkus/pull/12825/commits/62ee0faaf5105ab6a7e993c8be29bee2e49ee8fb to this PR after local testing passed and thought we could get this in at the same time. Let me know if you prefer it separated.

**Comment by Sanne** [comment]
looks good @machi1990 !  let's see CI output.

File: extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java [hunk 1]
```diff
@@ -86,6 +78,16 @@ NativeImageEnableAllTimeZonesBuildItem enableAllTimeZones() {
         return new NativeImageEnableAllTimeZonesBuildItem();
     }
 
+    @BuildStep
+    NativeImageSystemPropertyBuildItem disableAbandonedConnectionCleanUpInNativeMode() {
+        return new NativeImageSystemPropertyBuildItem("com.mysql.cj.disableAbandonedConnectionCleanup", "true");
+    }
+
+    @BuildStep
```
Comment (gsmet) [inline extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java:86]:
```suggestion
    @BuildStep(onlyIfNot = NativeBuild.class)
```
would make it clearer.

Comment (Sanne) [reply extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java:86]: I was wondering about that.. woudldn't it be enough to just have this single buildStep?

Comment (machi1990) [reply extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java:86]: > I was wondering about that.. woudldn't it be enough to just have this single buildStep?

Would `SystemPropertyBuildItem` be consumed and produce the `NativeImagePropertyBuildItem`? My first time producing either of them so I might be missing something.

Comment (Sanne) [reply extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java:86]: I don't know :) @gsmet ?

Comment (gsmet) [reply extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java]: I think it's OK the way it is now.

Comment (machi1990) [reply extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java]: > > I was wondering about that.. woudldn't it be enough to just have this single buildStep?

Tested it and it works by only producing the `SystemPropertyBuildItem`. But to get an answer to my question below ... 

> Would `SystemPropertyBuildItem` be consumed and produce the `NativeImageSystemPropertyBuildItem`? My first time producing either of them so I might be missing something.

With only the `SystemPropertyBuildItem` the command to generate native image is:

```bash
native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-DCoordinatorEnvironmentBean.transactionStatusManagerEnable=false -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Duser.language=en -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -jar quarkus-integration-test-jpa-mysql-999-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:+AddAllCharsets -H:EnableURLProtocols=http,https --enable-all-security-services -H:NativeLinkerOption=-no-pie --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-integration-test-jpa-mysql-999-SNAPSHOT-runner
```

With the `NativeImageSystemPropertyBuildItem` the command become:
```bash
native-image -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dcom.mysql.cj.disableAbandonedConnectionCleanup=true -J-DCoordinatorEnvironmentBean.transactionStatusManagerEnable=false -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Duser.language=en -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -jar quarkus-integration-test-jpa-mysql-999-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:+AddAllCharsets -H:EnableURLProtocols=http,https --enable-all-security-services -H:NativeLinkerOption=-no-pie --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-integration-test-jpa-mysql-999-SNAPSHOT-runner
```

Notice the `-J-Dcom.mysql.cj.disableAbandonedConnectionCleanup=true` in the second command which is missing from the first.

I think for the sake of being explicit in the generated command, I'd stick with producing the two build items.

Comment (Sanne) [reply extensions/jdbc/jdbc-mysql/deployment/src/main/java/io/quarkus/jdbc/mysql/deployment/JDBCMySQLProcessor.java]: Sure. But thanks for testing it! Good to know.

**Pull request merged**